File: solutions/200. Number of Islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        
        if(grid.empty()) {
            return 0;
        }
        
        int H = grid.size(),
            W = grid[0].size();
        
        int ans = 0;
        for(int row = 0; row < H; row++) {
            for(int col = 0; col < W; col++) {
                if(grid[row][col] == '1') {
                    ans += dfs(grid, row, col, H, W);
                }
            }
        }
        return ans;
    }
    
    int dfs(vector<vector<char>>& grid, int row, int col, int H, int W) {
        
        if(row < 0 or row >= H || col < 0 or col >= W || grid[row][col] != '1' ) {
            return 0;
        }
        
        grid[row][col] = '0';
        
        dfs(grid, row+1, col, H, W);
        dfs(grid, row-1, col, H, W);
        dfs(grid, row, col+1, H, W);
        dfs(grid, row, col-1, H, W);
        
        return 1;
    }
};

File: solutions/200. Number of Islands.java
class Solution {
        static int[] xdir = {-1, 0, 1, 0};
    static int[] ydir = {0, -1, 0, 1};
​
    // leetcode 200. https://leetcode.com/problems/number-of-islands/
    private static void islands_dfs(char[][] graph, int x, int y) {
        graph[x][y] = '0';
        for(int d = 0; d < 4; d++) {
            int r = x + xdir[d];
            int c = y + ydir[d];
            if(r >= 0 && r < graph.length && c >= 0 && c < graph[0].length && graph[r][c] != '0') {
                islands_dfs(graph, r, c);
            }
        }
    }
​
    public int numIslands(char[][] grid) {
        int islands = 0;
        for(int i = 0; i < grid.length; i++) {
            for(int j = 0; j < grid[0].length; j++) {
                if(grid[i][j] == '1') {
                    islands_dfs(grid, i, j);
                    islands++;
                }
            }
        }
        return islands;
    }
}
